intel_extension_for_transformers/llm/library/jblas/jblas/kernel_jit.h

```diff
@@ -740,6 +743,430 @@ static inline JBLAS_CODE decompress_s4_s8(utils::int4x2* srcptr, int8_t* dstptr,
+
+    Xbyak::Label l_idx_interleave_self;
+    std::shared_ptr<void> epilogue{
+        // generate code at the very end
+        nullptr, [&](void*) {
+          align(64);
+          L(l_idx_interleave_self);
+          db(reinterpret_cast<const uint8_t*>(idx_interleave_self), sizeof(idx_interleave_self));
+          outLocalLabel();  // end of local label
+
+          this->ready();
+          this->mKernel = this->getCode<func_t>();
+        }};
+    Xbyak::util::StackFrame st(this, 1, 13, 16 * 10 + SF_TmpSize);
+    const Xbyak::Reg64& parambase = st.p[0];
+    const Xbyak::Reg64& reg_srcptr = st.t[0];
+    const Xbyak::Reg64& reg_dstptr = st.t[1];
+    const Xbyak::Reg64& reg_srcstride = st.t[2];
+    const Xbyak::Reg64& reg_dststride = st.t[3];
+    const Xbyak::Reg64& reg_colsize = st.t[5];
+    const Xbyak::Reg64& reg_iterrow = st.t[6];
+    const Xbyak::Reg64& reg_itercol = st.t[7];
+    const Xbyak::Reg64& reg_tmp = st.t[8];
+    const Xbyak::Reg64& reg_tmp1 = st.t[9];
+    const Xbyak::Reg64& reg_tmp2 = st.t[12];
+    const Xbyak::Reg64& reg_tmp3 = st.t[10];
+
+    const Xbyak::Reg64& reg_ret = rax;
+    auto& mask_rd = k1;
+    const Xbyak::Zmm& vreg_idx0 = zmm31;
+
+    vreg_push(rsp);
+    vmovups(vreg_idx0, zword[rip + l_idx_interleave_self]);
+    mov(reg_srcptr, ptr[parambase + OFFSET(srcptr)]);
+    mov(reg_dstptr, ptr[parambase + OFFSET(dstptr)]);
+    mov(reg_srcstride.cvt32(), ptr[parambase + OFFSET(srcstride)]);
+    mov(reg_dststride.cvt32(), ptr[parambase + OFFSET(dststride)]);
+    mov(reg_colsize.cvt32(), ptr[parambase + OFFSET(col)]);
+
+    std::vector<Xbyak::Zmm> reg_srcs(row_pack), reg_tmps(row_pack);
+    const int ZIDX_TranSrc = 0;
+    const int ZIDX_TransTmp = row_pack;
+    for (int i = 0; i < row_pack; i++) reg_srcs[i] = Xbyak::Zmm(ZIDX_TranSrc + i);
+    for (int i = 0; i < row_pack; i++) reg_tmps[i] = Xbyak::Zmm(ZIDX_TransTmp + i);
+
+    xor_(reg_iterrow, reg_iterrow);
+    L(".rowloop");
+    xor_(reg_itercol, reg_itercol);
+    mov(reg_tmp2.cvt32(), ptr[parambase + OFFSET(row)]);
+    sub(reg_tmp2, reg_iterrow);
+    cmp(reg_tmp2, row_pack);
+    jb(".tailrowloop", T_NEAR);
+
+    L(".colloop");
+    mov(reg_tmp1, reg_itercol);
+    imul(reg_tmp1, reg_dststride);
+    lea(reg_tmp, ptr[reg_dstptr + reg_tmp1]);
+    lea(reg_tmp1, ptr[reg_srcptr + reg_itercol * src_bytes]);
+    for (int jj = 0; jj < n_tile; jj += ne_zmm) {
+      generate_Nbitsmask(mask_rd, reg_itercol, ptr[reg_colsize - jj], reg_tmp2, reg_tmp3, ne_zmm);
+      for (int ii = 0; ii < row_pack; ii++) {
+        const Xbyak::Xmm reg_srcs_ii = src_bytes_vmm == 64   ? Xbyak::Zmm(reg_srcs[ii].getIdx())
+                                       : src_bytes_vmm == 32 ? Xbyak::Ymm(reg_srcs[ii].getIdx())
+                                       : src_bytes_vmm == 16 ? Xbyak::Xmm(reg_srcs[ii].getIdx())
+                                                             : (assert(false), reg_srcs[ii]);
+        if (src_bytes == 1) {
```

Comment:
constexpr

Reply:
I tried to avoid the use of template in code generators to potentially save some time for compilation. It should be harmless to add some runtime branches in the logic of JIT code generation. As a result, it is not a constexpr here.